csrc/index_compute.cpp

```diff
@@ -3696,38 +3675,112 @@ std::pair<Val*, Val*> Index::getCpAsyncBulkGmemIndex(
 
   int64_t itemsize = dataTypeSize(gmem_tv->dtype());
 
-  for (auto it = frontier.rbegin(); it != frontier.rend(); it++) {
-    auto id = std::get<0>(*it);
-    tensor_sizes_inner_to_outer.push_back(id->extent());
-    if (it != frontier.rbegin()) {
-      tensor_strides_inner_to_outer.push_back(
-          SimplifyingIrBuilder::mulExpr(std::get<2>(*it), itemsize));
-    }
-    box_sizes_inner_to_outer.push_back(
-        partitioned_id_to_box_id.at(id)->extent());
-
-    auto stride_it = partitioned_id_to_stride_id.find(id);
-    if (it == frontier.rbegin()) {
-      NVF_ERROR(
-          stride_it == partitioned_id_to_stride_id.end(),
-          "When interleave is CU_TENSOR_MAP_INTERLEAVE_NONE ",
-          "(this is always the case for nvFuser now)",
-          ", the first element of elementStrides must be one.");
-    }
-    if (stride_it != partitioned_id_to_stride_id.end()) {
-      element_strides_inner_to_outer.push_back(stride_it->second->extent());
-    } else {
-      element_strides_inner_to_outer.push_back(gmem_tv->fusion()->oneVal());
-    }
+  // So far, we have infered the TMA domain. The size of TMA domain is not
+  // necessarily the dimensionality of TMA because we support defining box
+  // by compositing. We use a state machine to infer the dimensions of TMA.
+  //
+  // There can only be four types of IterDomains in the TMA domain:
+  // -  P: partitioned IterDomain
+  // -  C: coordinate IterDomain
+  // - SB: strided box IterDomain
+  // - CB: contiguous box IterDomain
+  //
+  // For the example of the Figure 6 in doc/dev/tma.md, the TMA domain is
+  // [I1, I2, I3, I4, I5, I6, I7, I8, I9], and the types of these IDs are
+  // [ C,  B,  P,  C,  B,  B,  C,  B,  B]
+  //
+  // The algorithm works as follows: We run a 3-state machine. The state machine
+  // is initialized as START. After setting the initial state, we loop through
+  // the TMA domain from inner to outer. During the loop, for each IterDomain we
+  // see, we take an action and change the state of the machine. The action and
+  // target state depend on the current state of the machine, and the type and
+  // contiguity of the IterDomain we encounter. The actions and transition of
+  // states are shown in the following diagram:
+  //
```

Comment:
Yes, in the example https://github.com/NVIDIA/Fuser/pull/1991#discussion_r1574259005, after visiting `I3`, the machine will be in state `START`. Then when visiting `I2`, it will hit `C`. Note that getting a `C` in `START` always indicate that there is a new dimension which has implicit box size 1.

Reply:
Oh, I see, so it's for an implicit size-one box.